internal/controller/mandatory_modules_deletion_controller.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	k8slabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/kyma-project/lifecycle-manager/api/shared"
+	"github.com/kyma-project/lifecycle-manager/api/v1beta2"
+	"github.com/kyma-project/lifecycle-manager/pkg/adapter"
+	"github.com/kyma-project/lifecycle-manager/pkg/log"
+	"github.com/kyma-project/lifecycle-manager/pkg/queue"
+	"github.com/kyma-project/lifecycle-manager/pkg/util"
+)
+
+const (
+	warningEvent          = "Warning"
+	settingFinalizerError = "SettingMandatoryModuleTemplateFinalizerError"
+	deletingManifestError = "DeletingMandatoryModuleManifestError"
+)
+
+type MandatoryModuleDeletionReconciler struct {
+	client.Client
+	record.EventRecorder
+	queue.RequeueIntervals
+}
+
+func (r *MandatoryModuleDeletionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logf.FromContext(ctx)
+	logger.V(log.DebugLevel).Info("Mandatory Module Deletion  Reconciliation started")
+
+	ctx = adapter.ContextWithRecorder(ctx, r.EventRecorder)
+
+	template := &v1beta2.ModuleTemplate{}
+	if err := r.Get(ctx, req.NamespacedName, template); err != nil {
+		if !util.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("MandatoryModuleDeletionController: %w", err)
+		}
+		logger.V(log.DebugLevel).Info(fmt.Sprintf("ModuleTemplate %s not found, probably already deleted",
+			req.NamespacedName))
+		return ctrl.Result{Requeue: false}, nil
+	}
+
+	if !template.Spec.Mandatory {
+		logger.V(log.DebugLevel).Info("Skipping reconciliation, ModuleTemplate not marked as mandatory")
+		return ctrl.Result{}, nil
```

Comment:
A question: When this item will be requeued? 
Maybe we can reduce the queue rate for non-mandatory modules by explicitly setting the re-queue time to some big interval, like:
`ctrl.Result{RequeueAfter: time.Hour*12}` ?
We can do that if we don't expect a non-mandatory module to become mandatory.
Is it a valid scenario to switch module templates between `mandatory` and `non-mandatory` in KCP?

Reply:
It will not be requeued at all, only if an event is happening on a ModuleController (create/update/delete) as usual in controllers. Thus I would NOT set `RequeueAfter` to 12 hours.

Please have a look at the `(r *MandatoryModuleDeletionReconciler) SetupWithManager` function, on how this reconciler is set up